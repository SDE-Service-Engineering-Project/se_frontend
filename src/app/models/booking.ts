import { Car } from './car';

export interface Booking {
  bookingId?: number;
  createdOn: string;
  bookedFrom: string;
  bookedUntil: string;
  bookingStatus?: string;
  price: number;
  currency: string;
  carId: string;
  userId?: number; // for now optional since we do not have a user id available
  daysToRent?: number;
  car: Car;
}
