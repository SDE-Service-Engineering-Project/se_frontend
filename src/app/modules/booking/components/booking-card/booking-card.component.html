<div class="card mb-2 mt-1" style="max-width: 540px">
  <div class="row no-gutters justify-content-center">
    <div class="col-3 d-flex">
      <img
        *ngIf="booking.bookingStatus === 'EXPIRED'"
        src="../../../../../assets/expired.png"
        class="card-img"
        style="margin: auto"
        alt="car"
      />

      <img
        *ngIf="booking.bookingStatus === 'BOOKED'"
        src="../../../../../assets/booked.png"
        class="card-img"
        style="margin: auto"
        alt="car"
      />
    </div>
    <div class="col-md-8">
      <div class="card-body">
        <h5 class="card-title">
          Vehicle:
          {{ booking.car && booking.car.brand ? booking.car.brand : "" }}
          {{ booking.car && booking.car.model ? booking.car.model : "" }}
        </h5>
        <hr />

        <p class="card-text mt-2">
          {{ booking.bookedFrom | date : "longDate" }} -
          {{ booking.bookedUntil | date : "longDate" }}
        </p>
        <p class="card-text">
          <span class="fw-bold">Booking days:</span>
          {{ getTimeDuration() | number : "1.0-0" }}
        </p>
        <hr />
        <p class="card-text mt-0">
          <span class="fw-bold">Total amount:</span>
          {{ booking.price | currency : booking.currency }}
        </p>
        <div *ngIf="booking.bookingStatus !== 'EXPIRED'">
          <button data-testid="cancel-booking-btn" class="btn btn-danger" (click)="cancelBooking(booking)">
            {{
              booking.bookingStatus === "BOOKED"
                ? "Cancel Booking"
                : "Withdraw Booking"
            }}
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
