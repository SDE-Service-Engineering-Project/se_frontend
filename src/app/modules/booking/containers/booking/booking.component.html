<div class="container" style="margin-top: 1rem">

  <div class="row">
    <h3>Current bookings</h3>
    <div class="col-6 justify-content-center"
         *ngFor="let booking of filterBookings('BOOKED')">
      <app-booking-card style="margin: 1rem" [booking]="booking"></app-booking-card>
    </div>
  </div>

  <div class="row">
    <h3>Expired Bookings</h3>
    <div class="col-6 justify-content-center" *ngFor="let booking of filterBookings('EXPIRED')">
      <app-booking-card style="margin: 1rem" [booking]="booking"></app-booking-card>
    </div>
  </div>


  <div class="container mt-1" *ngIf="bookings.length === 0">
    <div class="row">
      <h1 class="text-center">No Bookings found</h1>
    </div>
  </div>
</div>
