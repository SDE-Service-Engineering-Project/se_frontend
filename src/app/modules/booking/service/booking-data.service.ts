import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { Booking } from '../../../models/booking';

export const BOOKING_API = 'http://localhost:8080/api/v1/bookings';

@Injectable({
  providedIn: 'root',
})
export class BookingDataService {
  constructor(private http: HttpClient) {}

  fetchBookings(): Observable<Booking[]> {
    return this.http.get<Booking[]>(BOOKING_API);
  }
}
