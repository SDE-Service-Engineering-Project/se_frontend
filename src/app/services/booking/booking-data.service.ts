import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable, of } from 'rxjs';
import { Booking } from '../../models/booking';
import { environment } from '../../../environments/environment';

export const BOOKING_API = `${environment.baseUrl}bookings/`;

@Injectable({
  providedIn: 'root',
})
export class BookingDataService {
  constructor(private http: HttpClient) {}

  fetchBookings(): Observable<Booking[]> {
    return this.http.get<Booking[]>(BOOKING_API);
  }

  createBooking(booking: Booking): Observable<Booking> {
    return this.http.post<Booking>(BOOKING_API, booking);
  }

  cancelBooking(booking: Booking): Observable<any> {
    return of(['Booking was canceled']);
  }
}
