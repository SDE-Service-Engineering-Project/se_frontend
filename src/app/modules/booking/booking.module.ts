import { NgModule } from '@angular/core';
import { AsyncPipe, CommonModule } from '@angular/common';
import { BookingComponent } from './containers/booking/booking.component';
import { BookingCardModule } from './components/booking-card.module';
import { BookingRoutingModule } from './config/booking.routing';

@NgModule({
  declarations: [BookingComponent],
  imports: [BookingCardModule, AsyncPipe, CommonModule, BookingRoutingModule],
})
export class BookingModule {}
