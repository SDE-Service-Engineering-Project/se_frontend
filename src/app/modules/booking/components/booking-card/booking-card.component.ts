import { Component, EventEmitter, Input, Output } from '@angular/core';
import { Booking } from '../../../../models/booking';
import { NgbModal, NgbModalRef } from '@ng-bootstrap/ng-bootstrap';
import { ModalComponent } from '../modal/modal.component';
import { ToastService } from '../../../../services/toast/toast.service';

@Component({
  selector: 'app-booking-card',
  templateUrl: './booking-card.component.html',
  styleUrls: ['./booking-card.component.sass'],
})
export class BookingCardComponent {
  @Input() booking: Booking = {} as Booking;
  @Output() bookingCanceled: EventEmitter<any> = new EventEmitter<any>();

  modalRef!: NgbModalRef;

  constructor(
    private modalService: NgbModal,
    private toastService: ToastService
  ) {}

  getTimeDuration(): number {
    if (this.booking.bookedFrom && this.booking.bookedUntil) {
      return (
        (new Date(this.booking.bookedUntil).getTime() -
          new Date(this.booking.bookedFrom).getTime()) /
        (1000 * 3600 * 24)
      );
    }
    return 0;
  }

  cancelBooking(booking: Booking) {
    this.modalRef = this.modalService.open(ModalComponent);
    this.modalRef.result
      .then((decision) => {
        if (decision === 'confirmed') {
          this.bookingCanceled.emit(booking);
        }
      })
      .catch((err) => {
        this.toastService.showDefaultErrorToast(
          'Your Booking was not canceled'
        );
      });
  }
}
