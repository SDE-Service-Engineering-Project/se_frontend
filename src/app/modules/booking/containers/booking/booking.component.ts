import { Component } from '@angular/core';
import { Booking } from '../../../../models/booking';
import { BookingDataService } from '../../../../services/booking/booking-data.service';
import { ToastService } from '../../../../services/toast/toast.service';

@Component({
  selector: 'app-booking',
  templateUrl: './booking.component.html',
  styleUrls: ['./booking.component.sass'],
})
export class BookingComponent {
  bookings: Booking[];

  constructor(
    private bookingDataService: BookingDataService,
    private toastService: ToastService
  ) {
    this.bookings = [];
    this.bookingDataService.fetchBookings().subscribe({
      next: (data) => {
        this.bookings = data;
      },
      error: () => {
        this.toastService.showDefaultErrorToast(
          'Bookings are not available right now.'
        );
      },
    });
  }

  public filterBookings(type: string): Booking[] {
    return this.bookings?.filter((x) => x.bookingStatus === type);
  }

  //TODO: adapt to real response!
  cancelBooking(booking: Booking) {
    this.bookingDataService.cancelBooking(booking).subscribe({
      next: (response) => {
        this.toastService.showDefaultSuccessToast(
          'Your Booking was successfully canceled'
        );
      },
      error: (err) => this.toastService.showDefaultErrorToast(err),
    });
  }
}
