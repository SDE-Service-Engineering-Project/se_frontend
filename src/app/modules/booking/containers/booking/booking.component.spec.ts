import { ComponentFixture } from '@angular/core/testing';
import { createComponentFactory, Spectator } from '@ngneat/spectator/jest';
import { Component } from '@angular/core';
import { of } from 'rxjs';
import {
  HttpClientTestingModule,
  HttpTestingController,
} from '@angular/common/http/testing';
import { BookingComponent } from './booking.component';
import { BookingCardModule } from '../../components/booking-card.module';
import { mockBookings } from '../../../../utils/testing/mocks/booking.mock';
import { BookingDataService } from '../../service/booking-data.service';

@Component({
  template: '',
})
class RoutedTestComponent {}

describe('BookingComponent', () => {
  let spectator: Spectator<BookingComponent>;
  let component: BookingComponent;
  let fixture: ComponentFixture<BookingComponent>;
  let httpTestingController: HttpTestingController;

  const createComponent = createComponentFactory({
    component: BookingComponent,
    imports: [BookingCardModule, HttpClientTestingModule],
    providers: [BookingDataService],
  });

  beforeEach(() => {
    spectator = createComponent();
    httpTestingController = spectator.inject(HttpTestingController);
    component = spectator.component;
    fixture = spectator.fixture;
  });

  beforeEach(() => {
    component.bookings$ = of(mockBookings);
    spectator.detectChanges();
  });

  it('should render the component', () => {
    expect(fixture.debugElement.nativeElement).toMatchSnapshot();
  });
});
