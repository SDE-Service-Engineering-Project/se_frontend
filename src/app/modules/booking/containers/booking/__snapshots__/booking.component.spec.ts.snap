// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BookingComponent should render the component 1`] = `
<div
  id="root0"
>
  <div
    class="container"
    style="margin-top: 1rem;"
  >
    <div
      class="row"
    >
      <div
        class="col"
      >
        <app-booking-card
          style="margin: 1rem;"
        >
          <div
            class="card mb-3 mt-3"
            style="max-width: 540px;"
          >
            <div
              class="row no-gutters"
            >
              <div
                class="col-md-4"
              >
                <img
                  alt="car"
                  class="card-img"
                  src="https://external-content.duckduckgo.com/iu/?u=https%3A%2F%2Ftse1.mm.bing.net%2Fth%3Fid%3DOIP.GDp9uheqSrMQBmNhBUP1mwHaE0%26pid%3DApi&f=1&ipt=ee353c7a978b7ca256f57cae6ae51c4d5ca321133e806df55516214bfb5ba431&ipo=images"
                />
              </div>
              <div
                class="col-md-8"
              >
                <div
                  class="card-body"
                >
                  <h5
                    class="card-title"
                  >
                    Card ID: 1
                  </h5>
                  <p
                    class="card-text"
                  >
                    Booking: Mar 19, 2023 to Mar 19, 2023 
                  </p>
                  <p
                    class="card-text"
                  >
                    Total price: â‚¬12,345.00 
                  </p>
                </div>
              </div>
            </div>
          </div>
        </app-booking-card>
      </div>
      
      
    </div>
  </div>
</div>
`;
